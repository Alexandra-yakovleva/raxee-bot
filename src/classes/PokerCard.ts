export interface PokerCardRaw {
  suit: number
  value: number
}

export class PokerCard {
  constructor(public suit: number, public value: number) {}

  static fromRaw(raw: PokerCardRaw) {
    return new PokerCard(raw.suit, raw.value);
  }

  toRaw(): PokerCardRaw {
    return {
      suit: this.suit,
      value: this.value,
    };
  }

  toString() {
    return [
      ['♠️', '♣️', '♥️', '♦️'][this.suit],
      ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A'][this.value],
    ].join('');
  }
}
