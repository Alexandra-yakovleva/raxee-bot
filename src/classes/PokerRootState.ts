import { CustomContext } from '../types/context';

export interface PokerLobby {
  id: number
  userIds: number[]
}

export interface PokerRootStateRaw {
  lobbies: Record<string, PokerLobby>;
}

export class PokerRootState {
  lobbies: Record<string, PokerLobby> = {};

  constructor(public ctx: CustomContext) {}

  static fromRaw(ctx: CustomContext, raw: PokerRootStateRaw) {
    const instance = new PokerRootState(ctx);
    instance.lobbies = raw.lobbies;
    return instance;
  }

  toRaw(): PokerRootStateRaw {
    return {
      lobbies: this.lobbies,
    };
  }

  get lobbyByGroup(): PokerLobby | undefined {
    if (this.ctx.chat!.type === 'private') {
      throw new Error('lobbyByGroup called in private chat');
    }

    return this.lobbies[this.ctx.chat!.id];
  }

  get lobbyByUser(): PokerLobby | undefined {
    return Object.values(this.lobbies).find((lobby) => lobby.userIds.includes(this.ctx.from!.id));
  }

  get lobby(): PokerLobby | undefined {
    return this.ctx.chat!.type === 'private' ? this.lobbyByUser : this.lobbyByGroup;
  }

  addUserToLobby() {
    if (this.ctx.chat!.type === 'private') {
      throw new Error('addUserToLobby called in private chat');
    }

    if (!this.lobbies[this.ctx.chat!.id]) {
      this.lobbies[this.ctx.chat!.id] = {
        id: this.ctx.chat!.id,
        userIds: [],
      };
    }

    this.lobbies[this.ctx.chat!.id].userIds.push(this.ctx.from!.id);
  }

  removeLobby() {
    if (this.lobby) {
      delete this.lobbies[this.lobby.id];
    }
  }
}
