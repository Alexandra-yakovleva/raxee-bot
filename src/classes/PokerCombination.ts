import { PokerCard } from './PokerCard';

export class PokerCombination {
  constructor(public level: number, public cards: PokerCard[]) {}

  get weight() {
    return this.level * 1e10
    + (this.cards[0]?.value || 0) * 1e8
    + (this.cards[1]?.value || 0) * 1e6
    + (this.cards[2]?.value || 0) * 1e4
    + (this.cards[3]?.value || 0) * 1e2
    + (this.cards[4]?.value || 0);
  }

  get levelName() {
    return ['старшая карта', 'пара', 'две пары', 'тройка', 'стрит', 'флэш', 'фулл-хауз', 'каре', 'стрит флэш', 'флэш рояль'][this.level];
  }

  toString() {
    return [
      this.cards.join(' '),
      `(${this.levelName})`,
    ].join(' ');
  }
}
