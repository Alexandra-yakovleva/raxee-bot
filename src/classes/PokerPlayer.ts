import { User } from 'grammy/out/platform.node';
import * as R from 'remeda';

import { pokerMessages, pokerStrings } from '../constants/poker';
import { CustomContext } from '../types/context';
import { differenceWith, getPokerCombinations } from '../utils';

import { PokerCard, PokerCardRaw } from './PokerCard';

export interface PokerPlayerRaw {
  balance: number;
  bet: number;
  cards: PokerCardRaw[];
  folded: boolean;
  lost: boolean;
  turnMade: boolean;
  user: User;
}

export class PokerPlayer {
  balance = 1000;

  bet = 0;

  cards: PokerCard[] = [];

  folded = false;

  lost = false;

  turnMade = false;

  constructor(public ctx: CustomContext, public user: User) {}

  static fromRaw(ctx: CustomContext, raw: PokerPlayerRaw) {
    const instance = new PokerPlayer(ctx, raw.user);
    instance.balance = raw.balance;
    instance.bet = raw.bet;
    instance.cards = raw.cards.map((rawCard) => PokerCard.fromRaw(rawCard));
    instance.folded = raw.folded;
    instance.lost = raw.lost;
    instance.turnMade = raw.turnMade;
    return instance;
  }

  toRaw(): PokerPlayerRaw {
    return {
      balance: this.balance,
      bet: this.bet,
      cards: this.cards.map((card) => card.toRaw()),
      folded: this.folded,
      lost: this.lost,
      turnMade: this.turnMade,
      user: this.user,
    };
  }

  get combinations() {
    return getPokerCombinations([...this.ctx.pokerState.cards, ...this.cards]);
  }

  get ownCombinations() {
    return differenceWith(this.combinations, this.ctx.pokerState.boardCombinations, R.equals);
  }

  get topCombination() {
    return this.ownCombinations[0];
  }

  get win() {
    return !this.folded && this.topCombination.weight === this.ctx.pokerState.topWeight;
  }

  get callAmount() {
    return this.ctx.pokerState.topBet - this.bet;
  }

  get canFold() {
    return !(this.ctx.pokerState.isAllIn && this.balance === 0);
  }

  get canCheck() {
    return this.callAmount === 0;
  }

  get canCall() {
    return this.callAmount > 0 && this.callAmount < this.balance;
  }

  get canAllIn() {
    return !this.ctx.pokerState.isAllIn || (!this.canCheck && !this.canCall);
  }

  get canRaise() {
    return !this.ctx.pokerState.isAllIn;
  }

  get keyboard(): string[][] {
    const keyboard: string[][] = [
      this.ctx.pokerState.cards.map((card, index) => (index < this.ctx.pokerState.cardsOpened ? card.toString() : ' ')),
      [`Ð‘Ð°Ð½Ðº: ${this.ctx.pokerState.bankAmount} ðŸª™`],
    ];

    if (!this.lost) {
      keyboard.push([...this.cards.map(String), `${this.balance} ðŸª™`]);
    }

    if (this === this.ctx.pokerState.activePlayer) {
      keyboard.push(
        [
          this.canFold && pokerStrings.fold,
          this.canCheck && pokerStrings.check,
          this.canCall && pokerStrings.call(this.callAmount),
          this.canAllIn && pokerStrings.allIn,
        ].filter(Boolean) as string[],
      );
    }

    return keyboard;
  }

  async sendMessage(message: string, withKeyboard = false) {
    await this.ctx.api.sendMessage(
      this.user.id,
      message,
      {
        parse_mode: 'Markdown',
        ...withKeyboard && { reply_markup: { keyboard: this.keyboard } },
      },
    );
  }

  async setKeyboard() {
    await this.sendMessage(
      this === this.ctx.pokerState.activePlayer
        ? pokerMessages._.yourTurn(this.ctx.pokerState.activePlayer)
        : pokerMessages._.userTurn(this.ctx.pokerState.activePlayer),
      true,
    );
  }

  async increaseBet(amount: number) {
    amount = Math.min(amount, this.balance);

    this.bet += amount;
    this.balance -= amount;
  }
}
