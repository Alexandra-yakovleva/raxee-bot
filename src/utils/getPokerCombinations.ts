import { PokerCard } from '../classes/PokerCard';
import { PokerCombination } from '../classes/pokerCombination';

import { getPokerFlush } from './getPokerFlush';
import { getPokerFourOfAKind } from './getPokerFourOfAKind';
import { getPokerFullHouse } from './getPokerFullHouse';
import { getPokerHighHand } from './getPokerHighHand';
import { getPokerOnePair } from './getPokerOnePair';
import { getPokerRoyalFlush } from './getPokerRoyalFlush';
import { getPokerStraight } from './getPokerStraight';
import { getPokerStraightFlush } from './getPokerStraightFlush';
import { getPokerThreeOfAKind } from './getPokerThreeOfAKind';
import { getPokerTwoPairs } from './getPokerTwoPairs';

export const getPokerCombinations = (cards: PokerCard[]): PokerCombination[] => {
  cards = [...cards].sort((a, b) => (b.value - a.value) * 100 + (b.suit - a.suit));

  const highHand = getPokerHighHand(cards);
  const onePair = getPokerOnePair(cards);
  const twoPairs = getPokerTwoPairs(onePair);
  const threeOfAKind = getPokerThreeOfAKind(cards);
  const straight = getPokerStraight(cards);
  const flush = getPokerFlush(cards);
  const fullHouse = getPokerFullHouse(onePair, threeOfAKind);
  const fourOfAKind = getPokerFourOfAKind(twoPairs);
  const straightFlush = getPokerStraightFlush(straight);
  const royalFlush = getPokerRoyalFlush(straightFlush);

  return [
    ...royalFlush,
    ...straightFlush,
    ...fourOfAKind,
    ...fullHouse,
    ...flush,
    ...straight,
    ...threeOfAKind,
    ...twoPairs,
    ...onePair,
    ...highHand,
  ];
};
