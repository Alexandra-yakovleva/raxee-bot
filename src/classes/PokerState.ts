import * as R from 'remeda';

import { pokerMessages, pokerStickers, pokerStrings } from '../constants/poker';
import { CustomContext } from '../types/context';
import { getPokerCombinations, getRandomItem, shuffleItems } from '../utils';

import { PokerCard, PokerCardRaw } from './PokerCard';
import { PokerPlayer, PokerPlayerRaw } from './PokerPlayer';

export interface PokerStateRaw {
  activePlayerIndex: number;
  cards: PokerCardRaw[];
  cardsOpened: number;
  firstPlayerIndex: number;
  players: PokerPlayerRaw[];
  round: number;
  started: boolean;
}

export class PokerState {
  activePlayerIndex = 0;

  cards: PokerCard[] = [];

  cardsOpened = 0;

  firstPlayerIndex = -1;

  players: PokerPlayer[] = [];

  round = -1;

  started = false;

  constructor(public ctx: CustomContext) {}

  static fromRaw(ctx: CustomContext, raw: PokerStateRaw) {
    const instance = new PokerState(ctx);
    instance.activePlayerIndex = raw.activePlayerIndex;
    instance.cards = raw.cards.map((rawCard) => PokerCard.fromRaw(rawCard));
    instance.cardsOpened = raw.cardsOpened;
    instance.firstPlayerIndex = raw.firstPlayerIndex;
    instance.players = raw.players.map((rawPlayer) => PokerPlayer.fromRaw(ctx, rawPlayer));
    instance.round = raw.round;
    instance.started = raw.started;
    return instance;
  }

  toRaw(): PokerStateRaw {
    return {
      activePlayerIndex: this.activePlayerIndex,
      cards: this.cards.map((card) => card.toRaw()),
      cardsOpened: this.cardsOpened,
      firstPlayerIndex: this.firstPlayerIndex,
      players: this.players.map((player) => player.toRaw()),
      round: this.round,
      started: this.started,
    };
  }

  getPlayerByUserId(userId: number) {
    return this.players.find((player) => player.user.id === userId);
  }

  get nonLostPlayers() {
    return this.players.filter((player) => !player.lost);
  }

  get activePlayer() {
    return this.players[this.activePlayerIndex];
  }

  get bankAmount() {
    return this.players.reduce((sum, player) => sum + player.bet, 0);
  }

  get baseBet() {
    return (Math.floor(this.round / 4) + 1) * 10;
  }

  get topBet() {
    return Math.max(...this.players.map((player) => player.bet));
  }

  get isAllIn() {
    return this.nonLostPlayers.some((player) => player.balance === 0);
  }

  get boardCombinations() {
    return getPokerCombinations(this.cards);
  }

  getNextPlayerIndex(index: number, checkBalance = false) {
    do {
      index += 1;
      index %= this.players.length;
    } while (this.players[index].lost || this.players[index].folded || (checkBalance && this.players[index].balance === 0));

    return index;
  }

  async broadcastMessage(message: string, players = this.players) {
    await Promise.all(players.map((player) => player.sendMessage(message)));
  }

  async broadcastPlayerMessage(player: PokerPlayer, message: string) {
    const fullMessage = pokerMessages._.playerMessage(player, message);
    await Promise.all(this.players.map(async (p) => {
      if (player !== p) {
        await p.sendMessage(fullMessage);
      }
    }));
  }

  async setKeyboards() {
    await Promise.all(this.players.map((player) => player.setKeyboard()));
  }

  async dealCards() {
    await this.broadcastMessage(pokerMessages._.whoInGame(this.nonLostPlayers));

    const deck = shuffleItems(R.range(0, 52)).map((fullValue) => new PokerCard(fullValue % 4, Math.floor(fullValue / 4)));

    this.started = true;
    this.round += 1;
    this.cards = deck.splice(0, 5);
    this.cardsOpened = 3;

    this.players.forEach((player) => {
      player.bet = 0;
      player.cards = deck.splice(0, 2);
      player.folded = false;
      player.turnMade = false;
    });

    this.firstPlayerIndex = this.getNextPlayerIndex(this.firstPlayerIndex);
    this.activePlayerIndex = this.getNextPlayerIndex(this.firstPlayerIndex);

    const big = this.players[this.firstPlayerIndex];
    const small = this.activePlayer;

    big.increaseBet(this.baseBet * 2);
    small.increaseBet(this.baseBet);

    await this.broadcastMessage(pokerMessages._.blinds(big, small));
    await this.setKeyboards();
  }

  async finishGame() {
    await this.broadcastMessage(pokerMessages._.gameFinished);

    await Promise.all(this.players.map(async (player) => {
      await this.ctx?.api.sendSticker(player.user.id, getRandomItem(pokerStickers), { reply_markup: { remove_keyboard: true } });
    }));

    this.ctx.pokerRootState.resetLobby();
    this.ctx.pokerState = new PokerState(this.ctx);
  }

  async finishRound() {
    const topWeight = Math.max(...this.nonLostPlayers.map((player) => (player.folded ? -1 : player.topCombination.weight)));
    const winners = this.nonLostPlayers.filter((player) => !player.folded && player.topCombination.weight === topWeight);

    winners.forEach((player) => {
      player.balance += this.bankAmount / winners.length;
    });

    await this.broadcastMessage(pokerMessages._.roundFinished(this.cards, this.nonLostPlayers, winners));

    this.players.forEach((player) => {
      if (player.balance === 0) {
        player.lost = true;
      }
    });

    if (this.nonLostPlayers.length < 2) {
      await this.finishGame();
    } else {
      await this.dealCards();
    }
  }

  async nextTurn() {
    this.activePlayer.turnMade = true;

    if (this.nonLostPlayers.filter((player) => !player.folded).length < 2) {
      await this.finishRound();
      return;
    }

    if (this.players.every((player) => player.folded || player.balance === 0 || (player.turnMade && player.bet === this.topBet))) {
      if (this.cardsOpened === 5 || this.isAllIn) {
        await this.finishRound();
        return;
      }

      this.players.forEach((player) => {
        player.turnMade = false;
      });

      this.cardsOpened += 1;
      this.activePlayerIndex = this.getNextPlayerIndex(this.firstPlayerIndex - 1);
    } else {
      this.activePlayerIndex = this.getNextPlayerIndex(this.activePlayerIndex);
    }

    await this.setKeyboards();
  }

  async handleMessage(player: PokerPlayer, message: string) {
    switch (message) {
      case pokerStrings.fold: {
        if (!this.activePlayer.canFold) {
          return pokerMessages.onMessage.foldIsNotAllowed;
        }

        this.activePlayer.folded = true;
        await this.broadcastPlayerMessage(player, message);
        await this.nextTurn();
        break;
      }

      case pokerStrings.check: {
        if (!this.activePlayer.canCheck) {
          return pokerMessages.onMessage.checkIsNotAllowed;
        }

        await this.broadcastPlayerMessage(player, message);
        await this.nextTurn();
        break;
      }

      case pokerStrings.call(this.activePlayer.callAmount): {
        if (!this.activePlayer.canCall) {
          return pokerMessages.onMessage.callIsNotAllowed;
        }

        this.activePlayer.increaseBet(this.activePlayer.callAmount);
        await this.broadcastPlayerMessage(player, message);
        await this.nextTurn();
        break;
      }

      case pokerStrings.allIn: {
        if (!this.activePlayer.canAllIn) {
          return pokerMessages.onMessage.allInIsNotAllowed;
        }

        this.activePlayer.increaseBet(this.activePlayer.balance);
        await this.broadcastPlayerMessage(player, message);
        await this.nextTurn();
        break;
      }

      default: {
        const betAmount = Number(message);

        if (!betAmount) {
          await this.broadcastPlayerMessage(player, message);
          return pokerMessages.onMessage.unknownCommand;
        }

        if (!this.activePlayer.canRaise) {
          return pokerMessages.onMessage.raiseIsNotAllowed;
        }

        if (betAmount >= this.activePlayer.balance) {
          return pokerMessages.onMessage.betTooBig;
        }

        if (betAmount < this.activePlayer.callAmount + this.baseBet) {
          return pokerMessages.onMessage.betTooSmall;
        }

        this.activePlayer.increaseBet(betAmount);
        await this.broadcastPlayerMessage(player, message);
        await this.nextTurn();
      }
    }
  }
}
