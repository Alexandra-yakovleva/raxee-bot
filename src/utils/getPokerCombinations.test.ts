import { PokerCard } from '../classes/PokerCard';

import { getPokerCombinations } from './getPokerCombinations';

test('should sort cards strictly not in place and return combinations', () => {
  const cards = [
    new PokerCard(3, 2),
    new PokerCard(1, 9),
    new PokerCard(0, 2),
    new PokerCard(1, 3),
    new PokerCard(2, 2),
    new PokerCard(2, 9),
    new PokerCard(3, 9),
  ];

  expect(getPokerCombinations(cards)).toMatchSnapshot();

  expect(cards).toStrictEqual([
    new PokerCard(3, 2),
    new PokerCard(1, 9),
    new PokerCard(0, 2),
    new PokerCard(1, 3),
    new PokerCard(2, 2),
    new PokerCard(2, 9),
    new PokerCard(3, 9),
  ]);
});
