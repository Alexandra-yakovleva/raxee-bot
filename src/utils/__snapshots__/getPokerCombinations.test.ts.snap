// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should sort cards strictly not in place and return combinations 1`] = `
Array [
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
    ],
    "level": 6,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 6,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 6,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
    ],
    "level": 6,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
    ],
    "level": 6,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
    ],
    "level": 6,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
    ],
    "level": 3,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 3,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 2,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 2,
        "value": 9,
      },
    ],
    "level": 1,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
    ],
    "level": 1,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 9,
      },
      PokerCard {
        "suit": 1,
        "value": 9,
      },
    ],
    "level": 1,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 2,
        "value": 2,
      },
    ],
    "level": 1,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 1,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 2,
      },
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 1,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 9,
      },
    ],
    "level": 0,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 9,
      },
    ],
    "level": 0,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 1,
        "value": 9,
      },
    ],
    "level": 0,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 1,
        "value": 3,
      },
    ],
    "level": 0,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 3,
        "value": 2,
      },
    ],
    "level": 0,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 2,
        "value": 2,
      },
    ],
    "level": 0,
  },
  PokerCombination {
    "cards": Array [
      PokerCard {
        "suit": 0,
        "value": 2,
      },
    ],
    "level": 0,
  },
]
`;
